File: cpp/dsa/algorithms/binary-search/koko.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1;
        int right = 0;
        for (int pile : piles) right = max(right, pile);
        int result = right;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            long long hours = 0;
            for (int pile : piles) {
                hours += (pile + mid - 1) / mid;
            }

            if (hours <= h) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return result;
    }
};

File: python/dsa/algorithms/binary-search/koko.py
def minEatingSpeed(piles, h):
    left, right = 1, max(piles)
    result = right

    while left <= right:
        mid = left + (right - left) // 2
        
        hours = 0
        for pile in piles:
            hours += (pile + mid - 1) // mid
            
        if hours <= h:
            result = mid
            right = mid - 1
        else:
            left = mid + 1
            
    return result